File: app/src/test/java/de/westnordost/osmagent/data/osm/tql/TestBooleanExpressionParser.java
package de.westnordost.osmagent.data.osm.tql;

public class TestBooleanExpressionParser
{
	public static BooleanExpression<BooleanExpressionValue> parse(String input)
	{
		BooleanExpressionBuilder<BooleanExpressionValue> builder = new BooleanExpressionBuilder<>();
		StringWithCursor reader = new StringWithCursor(input);
		while(!reader.isAtEnd())
		{
			if(reader.nextIsAndAdvance('*')) builder.addAnd();
			else if(reader.nextIsAndAdvance('+')) builder.addOr();
			else if(reader.nextIsAndAdvance('(')) builder.addOpenBracket();
			else if(reader.nextIsAndAdvance(')')) builder.addCloseBracket();
			else builder.addValue(new TestBooleanExpressionValue(String.valueOf(reader.advance())));
		}
		return builder.getResult();
	}
}

File: app/src/commonTest/kotlin/de/westnordost/streetcomplete/data/elementfilter/TestBooleanExpressionParser.kt
package de.westnordost.streetcomplete.data.elementfilter

import de.westnordost.streetcomplete.util.StringWithCursor

object TestBooleanExpressionParser {
    fun parse(input: String): BooleanExpression<Matcher<String>, String>? {
        val builder = BooleanExpressionBuilder<Matcher<String>, String>()
        val reader = StringWithCursor(input)
        while (!reader.isAtEnd()) {
            when {
                reader.nextIsAndAdvance('*') -> builder.addAnd()
                reader.nextIsAndAdvance('+') -> builder.addOr()
                reader.nextIsAndAdvance('(') -> builder.addOpenBracket()
                reader.nextIsAndAdvance(')') -> builder.addCloseBracket()
                else -> builder.addValue(TestBooleanExpressionValue(reader.advance().toString()))
            }
        }
        return builder.build()
    }
}
